ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp

```diff
@@ -351,6 +355,118 @@ const operation::Hash ScaledDotProductAttentionDecode::compute_program_hash(
+
+    // Check num_heads
+    TT_FATAL(k_shape[1] == v_shape[1]);
+    TT_FATAL(q_shape[1]%k_shape[1]==0);
+    TT_FATAL(q_shape[1]<=32);
+
+    // Check batch size
+    TT_FATAL(k_shape[0] == v_shape[0]);
+
+    // Check valid seqlen
+    for (int i = 0; i < this->cur_pos.size(); i++) {
+        TT_FATAL(this->cur_pos[i] < k_shape[-2], "cur_pos must be <= K sequence dim");
+    }
+
+    // Check compute kernel config
+    std::visit([&](auto&& compute_kernel_config) {
+            using T = std::decay_t<decltype(compute_kernel_config)>;
+            if constexpr (std::is_same_v<T, WormholeComputeKernelConfig>) {
+                TT_FATAL(compute_kernel_config.fp32_dest_acc_en == false, "FP32 dest acc disabled due to nd pcc and unpacker hang issue.");
+            }
+        }, this->compute_kernel_config);
+}
+
+std::vector<Shape> ScaledDotProductAttentionGQADecode::compute_output_shapes(const std::vector<Tensor>& input_tensors) const {
+    auto tt_q_shape = input_tensors.at(0).get_legacy_shape();
+    auto tt_k_shape = input_tensors.at(1).get_legacy_shape();
+    uint32_t n_groups = tt_q_shape[2]/tt_k_shape[1];
+    return {input_tensors.at(0).get_legacy_shape()};
+}
+std::vector<Tensor> ScaledDotProductAttentionGQADecode::create_output_tensors(const std::vector<Tensor>& input_tensors) const {
+    return operation::generic_create_output_tensors(
+        *this, input_tensors, input_tensors.at(0).get_dtype(), Layout::TILE, this->output_mem_config);
+}
+operation::ProgramWithCallbacks ScaledDotProductAttentionGQADecode::create_program(
+    const std::vector<Tensor>& input_tensors,
+    std::vector<Tensor>& output_tensors) const {
+    auto& input_tensor_q = input_tensors.at(0);
+    auto& input_tensor_k = input_tensors.at(1);
+    auto& input_tensor_v = input_tensors.at(2);
+    auto& output_tensor = output_tensors.at(0);
+
+    auto scale = this->scale;
+    if (not scale.has_value()) {
+        scale = 1.0f / std::sqrt(static_cast<float>(input_tensor_q.get_legacy_shape()[-1]));
+    }
+
+    // TODO: get this from program_config
+    std::size_t q_chunk_size;
+    std::size_t k_chunk_size;
+
+    std::visit(
+        [&](const auto& program_config) {
+            using ProgramConfigType = std::decay_t<decltype(program_config)>;
+            if constexpr (std::is_same_v<
+                              ProgramConfigType,
+                              tt::operations::primary::transformers::SDPAMultiCoreProgramConfig>) {
+                q_chunk_size = program_config.q_chunk_size;
+                k_chunk_size = program_config.k_chunk_size;
+            } else {
+                q_chunk_size = k_chunk_size = 32;
+            }
+        },
+        this->program_config);
+
+    return sdpa_decode_multi_core(
+        input_tensor_q,
+        input_tensor_k,
+        input_tensor_v,
+        output_tensor,
+        this->cur_pos,
+        scale,
+        this->compute_kernel_config,
+        this->program_config,
+        this->k_chunk_size);
```

Comment:
You don't even use `q_chunk_size` or `k_chunk_size` here, did you intend to remove these and the `std::visit`?